java/com/google/re2j/ProgBuilder.java

```diff
@@ -0,0 +1,118 @@
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Original Go source here:
+// http://code.google.com/p/go/source/browse/src/pkg/regexp/syntax/prog.go
+
+package com.google.re2j;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A ProgBuilder is a builder for a Prog.
+ */
+final class ProgBuilder {
```

Comment:
Again, why do we need two representations of Prog, one mutable, one not?  We could reimplement List.add easily enough.  Does making these three fields final really affect the running time?

Reply:
I did that first with copying the every time we added one. but then I created more garbage  everytime an inst was modified. I think there is a benefit of having a final version, as to garantee that once it's build it won't get modified. also the build function is then a good place to put optimisation as we now that after it won't change.

Reply:
saying that  I can remove the builders if it's a problem.